### ISSUE TITLE
Unable to handle inductive examples from ESOP-13

### ISSUE THREAD

**Issue by shingarov** [ranjitjhala/sprite-lang#4]
In an ideal world, Sprite is supposed to cover all of Vazou–Rondon–Jhala's ESOP-13 paper [Abstract Refinement Types](https://goto.ucsd.edu/~rjhala/papers/abstract_refinement_types.pdf).
But I am having difficulty with `foldr` on p.10:
```
foldr :: forall <p :: [a] -> b -> Bool>.
             (xs:[a] -> x:a -> b <p xs> -> <p (x:xs)>)
          -> b<p []>
          -> ys:[a]
          -> b<p ys>
foldr op b [] = b
foldr op b (x:xs) = op xs x (foldr op b xs)
```
This would naïvely transliterate into Sprite like this:
```
type list('a) =
  | Nil
  | Cons (x:'a, xs:list('a))
  ;

/*@ val foldr : forall (p: list('a) => 'b => bool).
                (xs:list('a) => x:'a => acc:'b[v|p xs v] => 'b[v|p (Cons x xs) v])  =>
                seed:'b[v|p (Nil) v]                                                =>
                ys:list('a)                                                         =>
                'b[v|p ys v]                  */
let rec foldr = (op, seed, ys) => {
  switch (ys) {
  | Nil        => seed
  | Cons(h, t) => let nextb = foldr(op, seed, t);
                  op(t, h, nextb)
  }
};
```
which is of course not even well-formed Sprite, because args in `($p ... )` can only be variables, not anywhich Exprs.  And for sure, the above crashes Sprite in `rvarArgSymbol`:
```
$ stack exec --profile -- sprite 6 listFoldr.re
Unexpected argument in ref-variable: p (Cons x xs) v EApp (EApp (EVar "Cons") (EVar "x")) (EVar "xs")
CallStack (from -prof):
  Language.Sprite.L6.Types.rvarArgSymbol (src/Language/Sprite/L6/Types.hs:(467,1)-(468,124))
  Language.Sprite.L6.Types.isRVarApp (src/Language/Sprite/L6/Types.hs:(461,1)-(464,36))
  Language.Fixpoint.Misc.mapEither (src/Language/Fixpoint/Misc.hs:(330,1)-(335,48))
  Language.Sprite.L6.Types.refactorAppP.(...) (src/Language/Sprite/L6/Types.hs:458:5-58)
  Language.Sprite.L6.Types.refactorAppP (src/Language/Sprite/L6/Types.hs:(455,1)-(458,58))
  Language.Sprite.L6.Types.refactorAppR (src/Language/Sprite/L6/Types.hs:(450,1)-(451,33))
  Language.Sprite.L6.Types.fmap (src/Language/Sprite/L6/Types.hs:57:23-29)
  Language.Sprite.L6.Types.refactorApp (src/Language/Sprite/L6/Types.hs:(428,1)-(432,30))
  Language.Sprite.L6.Types.generalize (src/Language/Sprite/L6/Types.hs:326:1-41)
  Language.Sprite.L6.Parse.mkDecl (src/Language/Sprite/L6/Parse.hs:(249,1)-(255,24))
```

This seems reasonable, so I went a bit further and tried to get around this limitation by introducing two additional ghosts, `nil:list('a)[v| nil = Nil]` and `cns:list('a)[v | cns = Cons(x,xs) ]` — the idea being that the offending applications `($p (Consx,xs) v)` would become `($p cns v)` — but this only kicks the can down the road because you can't use `(Cons x xs)` in a predicate like that either (so this crashes Sprite too).

So I went another step and tried to generalize `($p ... )`  to take arbitrary Exprs under the hypothesis that those `H.Var`s ultimately all become `PKVars`, like in https://github.com/shingarov/sprite-lang/commit/84d2ea6819ef128bbf640fbceb6f0e7f4c728f2e which doesn't work either because those `($p ... )`  `H.Var`s do NOT become `PKVar`s but get picked out into `$k##1` etc.

**Comment by ranjitjhala** [comment]
Thanks @shingarov ! Re-reading this text, I understand the issue now (I think!)

1.  The `Horn` constraint format in `Fixpoint` requires that all `Kvar` applications be to `Symbol` 

```haskell
data Pred
  = Reft  !F.Expr                               -- ^ r
  | Var   !F.Symbol ![F.Symbol]        -- ^ $k(y1..yn)
  | PAnd  ![Pred]                               -- ^ p1 /\ .../\ pn
  deriving (Data, Typeable, Generic, Eq)
```

2. **but this particular example** wants a KVar application to a term `Cons x xs`

I think the general solution here is to *generalize* the `Var` case for `Pred` so that it is

```haskell
data Pred = ... | Var !F.Symbol ![F.Expr]
```

This is a bit more generous for clients using the Horn format (like SPRITE, and also `flux`).

The _other_ option is to do what `flux` does currently which is have an IR that allows `Expr` as arguments,
but then *translate* them away by adding the extra ghost binders as you are attempting so, e.g.

```
... => ... K(e1,e2) ... => ...
```

would become

```
... => forall x1. x1 = e1 =>  forall x2. x2 = e2 => ... K(x1,x2) ... => ...
```

@nilehmann do you have some code in `flux` that systematically does this "ANF"-ing for Horn constraints? If so, maybe we can port it inside `liquid-fixpoint`?

**Comment by nilehmann** [comment]
This is the code that does the ANFing https://github.com/flux-rs/flux/blob/main/flux-refineck/src/fixpoint_encoding.rs#L291. Some things are very specific to flux, but it should be easy to implement the same directly in fixpoint. However, I wonder how dependent is fixpoint on kvars being only applied to variables? Does it need this constraint to do the search?

I mention this because an extension I would really like to see in fixpoint is to generalize the "self" argument to a list of "self-arguments". Right now, fixpoint only searches for solutions that use the first argument (which I'm calling the self-argument). I would like to extend this such that it searches for solutions that use at least one of a list of arguments. 

Since there's no currying in Rust, it often happens that I need a kvar on a list of arguments. I've been using the syntax `$k(x, y, z)[a, b, c]` where I call the arguments `x`, `y` and `z` the _self arguments_ and `a`, `b` and `c` the _scope_. I would like the solution of `$k` to use at least one of `x`, `y` or `z`, and it may or may not use `a`, `b` and `c`. Currently, the way I encode this is by generating a conjunction of kvars:

```
$k(x, y, z)[a, b, c] := $k1(z)[a, b, c] ∧ $k2(y, z)[a, b, c] ∧ $k3(x, y, z)[a, b, c]
```

I think the search would be more efficient if supported natively by fixpoint.

**Comment by shingarov** [comment]
I just tried to take another stab at it, and now the problem seems larger than I first thought.  The "plain abstract refinements" (or to be precise, in what ICFP-15 calls λP), not only can subexpressions appear inside κ-applications, but κ-applications themselves can be wrapped under logical connectives (not necessarily only ∧ as in @nilehmann's example).

Say, in the "function composition" example in ICFP-15, the λB
```
bound Chain p q r = λx y z →
  q x y ⇒ p y z ⇒ r x z
```
from p.3 gets translated to this λP on p.4:
```
type ChainTy p q r
  = x:a → w:b → z:c →
    {v: Bool | q x w ⇒ p w z ⇒ r x z}
compose :: ( ChainTy p q r) → (y:b → c <p y >)
  → (z:a → b <q z >)
  → (x:a → c <r x >)
compose $chain f g x =
  let t1 = g x
  t2 = f t1
  _ = $chain x t1 t2 -- materialize
  in t2
```
Now if we naïvely try to express this in Sprite:
```
/*@ val compose : forall (p : 'b => 'c => bool).
                  forall (q : 'a => 'b => bool).
                  forall (r : 'a => 'c => bool).
                  (x:'a => w:'b => z:'c => Bool[v| (q x w) ==> ((p w z) ==> (r x z)) ])
               => (y:'b => 'c[v | p y v])
               => (z:'a => 'b[v | q z v])
               =>  x:'a
               => 'c[v | r x v]                */
let compose = (cha, f, g, x) => {
  let t1 = g(x);
  let t2 = f(t1);
  let t3 = cha(x, t1, t2);
  t2
};
```
the problem is the `(q x w) ==> ((p w z) ==> (r x z))` which will obviously crash, see `[NOTE:RefactorApp]`: the result of `isRVarApp` is `Either` a (top-level!!!) κ-app, or an Expr not containing ref-vars.

**Comment by ranjitjhala** [comment]
Sorry for the delay with this; I got a bit sidetracked getting `sprite` up-to-date with the latest fixpoint (there are various little changes ...) Two things

1. This above example requires "bounds" which `sprite` doesn't support at all, but which would be nice to add.

2. The syntax `(q x w) ==> (p w z) => (r x z)` can be a bit misleading -- this is probably currently parsed directly as an `Expr` but in reality, it should be parsed as a Horn constraint i.e. `Horn.Cstr`. Recall that the grammar for Horn constraints is this:

https://github.com/ucsd-progsys/liquid-fixpoint/blob/develop/src/Language/Fixpoint/Horn/Types.hs#L168-L173

So really that implication -- when part of a "bound" -- should get parsed and represented as below

```
(forall (_ (int)) (q x w) 
  (forall (_ (int)) (p w z) 
     (r x z)
  )
)
```

**Comment by ranjitjhala** [comment]
So if we wanted to add `bound` to `sprite` we would require a way to add `H.Cstr` inside a signature. 

```haskell
data Type r 
  = TBase !Base                         r    -- ^ Int{r} 
  | TFun  !F.Symbol !(Type r) !(Type r)      -- ^ x:s -> t 
  | TAll  !TVar     !(Type r)                -- ^ all a. t
  | TCon  !TyCon    ![Type r]           r    -- ^ C t1...tn 
  | TBnd !TCstr !(Type r)
  deriving (Eq, Ord, Show)
```

where `TCstr` uses `Horn.Cstr` internally.

But may be easier to get the `foldr` example working via Nico's translation above, before adding `bound`.

**Comment by shingarov** [comment]
>getting sprite up-to-date with the latest fixpoint

Yes, I have been following what's going on in the `bump-fixpoint` branch :-)

>This above example requires "bounds" which sprite doesn't support at all, but which would be nice to add

Wait... there must be a confusion.  I do understand what you say in the above paragraph; that's what I started to do which triggered the whole original thread in slack.  The confusion is, when you say "requires bounds", this is about the λB (if we follow ICFP-15 terminology).  IIUC, the task of "an implementation of bounds" is translating that λB into λP.  So for myself, to get better clarity about how this translation will work, I wanted an example in λB with a corresponding manually-written example in λP.  Such explicit pair of examples is given by `Chain` and `ChainTy` in ICFP-15.  So before going to the next stage of actually attempting to write the "bounds translator", I wanted to have a (successfully typechecked) ChainTy in Sprite.  It appeared it should be easy, because (to the naïve reader) ICFP-15's λP is the same as Sprite's L₆ and the same as ESOP-13.  (In other words, ChainTy does NOT require Bounds, but the other way around: Bounds would require that its implementation substrate would already deal with ChainTy).  The crux of the present Issue is that λP ("plain" abstract refs) is NOT expressible strictly on top of L₆.

Am I making sense?

>So if we wanted to add `bound` to `sprite` we would require a way to add `H.Cstr` inside a signature.

Ah, I see — this would be consistent with the structure of the Tutorial, some "L₉" would extend L₆ by adding Bounds.  My original idea was to do Bounds strictly-on-top of Sprite.  I guess extending Sprite is a better style to do the same thing.

>But may be easier to get the foldr example working via Nico's translation above, before adding bound.

They are two different examples, but I think if we fix the underlying issue that L₆ is short of λP, then *both* problems go away, don't they?  Both that is, (1) ESOP-13's example works, and (2) `ChainTy` typechecks so one is unblocked to start adding Bound.

**Comment by ranjitjhala** [comment]
Ah, my apologies -- I had misunderstood your goal, upon reading more closely I follow now! 

So the issue is still this bit

```              
(x:'a => w:'b => z:'c => Bool[v| (q x w) ==> ((p w z) ==> (r x z)) ])
```

where the `q`, `p` appear in non-toplevel. Instead, if you **change** that line to the equivalent

```
(x:'a => w:'b[v|q x v] => z:'c[v|p w v] => Bool[v|r x z])
```

then in fact, the code works! (at least on ` bump-fixpoint` ... I am assuming also in yours?)

The more "direct" encoding of the "bound" from ICFP15 would be something like the below 
where each "antecedent" of the bound becomes an "argument"

```
(x:'a => w:'b => z:'c =>  int[v|q x w] => int[v|p w z] => Bool[v|r x z])
```

and then of course you change the line to

```  
let t3 = cha(x, t1, t2, 0, 0);
```

and that also works. 

Does that make sense?

**Comment by ranjitjhala** [comment]
PS: I added those to the `bump-fixpoint` branch so you can run them with

```
cabal run -- sprite 6 test/L6/pos/compose00.re
cabal run -- sprite 6 test/L6/pos/compose01.re
```

**Comment by shingarov** [comment]
>at least on  bump-fixpoint ... I am assuming also in yours?

Yes, both that and the "more direct" `compose0x` work.  Now everything makes perfect sense!!!

**Comment by ranjitjhala** [comment]
Dear @shingarov -- no worries, authors always like it when someone takes enough interest to ask such questions!
I'm traveling to India ATM so will take a while to page in the details - but did you see the paragraph "Rule CApp" at the bottom of p6 that discusses this rule?

**Comment by shingarov** [comment]
> the paragraph "Rule CApp" at the bottom of p6 that discusses this rule

Yes, that paragraph is exactly what I can't understand.  So we discard all the refinements — for learning sake let's say in the `ex2` example this yields just `true` — and then this resulting Const (i.e., the `true`) is somehow getting subtyped to carry the bound's refinement.  How does this subtyping work?  To me, "subtyping" suggests there would need to be a κ-variable somewhere in the translated λP.  Well, "translated λP" in this case cannot be anything other than "_e′ Const(φ)_"; and _Const(φ)_ is just `true` so the translated λP is "_e true_"? (because in this case _e⇝e′=e_).  I don't see how this makes sense.  What is the magic that introduces κ?  ok I could see how subtyping would lead to the correct tautology if we somehow got to the instantiation described on p.3 (left column, just above "List Filtering"): "*To verify* `ex2` we instantiate, at the call to `compose`,
```
p, q ↦ λx v → v = x + 1
   r ↦ λx v → v = x + 2
```

This comes from where — from CAbs???  No matter how hard I squint at it, I can't see how this can be.

As a learning aid, I have bit the bullet and started looking at how this is implemented in LiquidHaskell.  Ok so in `benchmarks/icfp15/pos/Composition.hs` and the others in that directory, we don't give names (as the paper does) such as "Chain" to bounds, instead we expand them in-line to
```
{x::a, w::b<q x> |- c<p w> <: c<r x>}
```
which the `constraintP` parser places into the left half of `RRTy` (i.e. the part which is on the left of the `=>`).  So that's the last line on p.6 (bottom of right column).  It seems pretty obvious, too, how this translates to the λP of `test/L6/pos/compose00.re` in Sprite.

So that's for `compose` (aka `cmp` in `Composition.hs` parlance).  Then when we call `compose` from `ex2` (`incr2` in `Composition.hs`), in λP we need to pass it the ghost function (`cha` in `compose00.re` parlance) which is just `λ_ _ _ → true`.  How is that latter typed?  In other words, what triggers the instantiation
```
p, q ↦ λx v → v = x + 1
   r ↦ λx v → v = x + 2
```
?  My understanding is that `compose(incr, incr)` is just a normal `EApp`, how does it decide to instantiate? and where in liquidhaskell-boot is that instantiation itself?

Perhaps if you could expand `compose00.re` with the λP-translated version of `ex2`, I might suddenly get enlightened by looking at it?

**Comment by ranjitjhala** [comment]
Yup, that's exactly the idea! 

Re: `Inst` / `Wrap` / `Instances` -- I'll have to dig around -- @nikivazou wrote that code IIRC, Niki, do you recall where in the LH code base you "instantiate" the bounds?

**Comment by shingarov** [comment]
Ping?

**Comment by ranjitjhala** [comment]
Oops forgot this — @nikivazou wrote this code so maybe she can find it
faster than me I’ll take a look after my lecture today!

- Ranjit.

**Comment by nikivazou** [comment]
To be honest, I do not remember all the details, but [instantiatePreds](https://github.com/ucsd-progsys/liquidhaskell/blob/develop/liquidhaskell-boot/src/Language/Haskell/Liquid/Constraint/Generate.hs#L415) is a good starting point to trace all the bounded refinement types code...

**Comment by shingarov** [comment]
I was content thinking that I understand how instantiation is done.
But now after letting this settle in my brain for quite some time, I went to think about this again, and now I am *truly* confused by what `Instances(Γ,Φ)` does as explained by p.6, left column, of the ICFP-15 paper.  How can it be that the LET translation rule is semantics-preserving?  How is it possible that it doesn't instantiate more constraints than is meaningful, if it simply produces all combinations that fit the unrefined (plain Hindley–Milner) shape?

For example, the polymorphic `compose` does not present a problem: when verifying
```

{-@
compose :: forall < p :: b -> c -> Bool
              , q :: a -> b -> Bool
              , r :: a -> c -> Bool
              >.
       {x::a, w::b<q x> |- c<p w> <: c<r x>}
       f:(y:b -> c<p y>)
    -> g:(z:a -> b<q z>)
    -> x:a -> c<r x>
@-}
compose :: (b -> c)
    -> (a -> b)
    ->  a -> c 
compose f g x = 
  let z1 = g x     in
  let z2 = f z1 in
  z2
```

there is only one way to materialize the application of `$chain` to the variables in scope — because their types `a`, `b`, `c` only fit `$chain`'s argument types in one way — so `Wrap(e,{e₁,…,eₙ})` gives only one
```
let t₁ = $chain(x, z1, z2) in …
```

However, consider what would happen if `compose` were monomorphic:
```
{-@
compose :: forall < p :: Int -> Int -> Bool
              , q :: Int -> Int -> Bool
              , r :: Int -> Int -> Bool
              >.
       {x::Int, w::Int<q x> |- Int<p w> <: Int<r x>}
       f:(y:Int -> Int<p y>)
    -> g:(z:Int -> Int<q z>)
    -> x:Int -> Int<r x>
@-}
compose :: (Int -> Int)
    -> (Int -> Int)
    ->  Int -> Int
compose f g x =
  let z1 = g x  in
  let z2 = f z1 in
  z2
```

I think this would be catastrophic, because there are many ways to apply `$chain` to `x`, `z1`, `z2` and the LET rule will give
```
let t₁ = $chain(x, z1, z2) in
let t₂ = $chain(x, z2, z1) in
…
```
and this λP does not typecheck.

Now.

To try to understand this, I started experimenting with how actual LiquidHaskell would handle the monomorphic `compose`.  Both the pos and neg cases give the expected result.  So I drilled down to `addLhsInv` in `Split.hs` and `constraintToLogic` in `Constraint.hs` where the combinations seem to be enumerated (is this even the correct place to look?? I am a bit scared by the vacillating comment at the top of `Constraint.hs`).  `constraintToLogic` produces a `pAnd` of about 100 combinations, and by looking at the code I am unable to see how it can tell apart "good" ones from such meaningless ones that would cause verification to fail.

Can you explain how this works?  I am at my wits' end.

**Comment by ranjitjhala** [comment]
Hi Boris,

The key bit here is that the “extra” instances, t1, t2 etc *do not* cause
the verification to fail. Each of these just adds extra facts to the
environment that is used to prove the output type for compose, so as long
as ONE OF the instances has the right fact, the code type checks. You are
right that in  the polymorphic version there is only ONE well typed
instance, and in the monomorphic version there are MANY (27 = 3^3) However
one of those 27 is exactly the same one as the polymorphic one and so the
code verifies. The others simply hang around as unused assumptions in the
environment (making checking slow, but not actually invalidating the
check.) Does that help?
- Ranjit.
